src/Features/Core/Portable/Diagnostics/DiagnosticCustomTags.cs

```diff
@@ -51,5 +53,28 @@ private static void Assert(string[] customTags, params string[] tags)
                 Debug.Assert(customTags[i] == tags[i]);
             }
         }
+
+        internal static string[] Create(bool isUnneccessary, bool isConfigurable, params string[] customTags)
+        {
+            if (customTags.Length == 0 && isConfigurable)
+            {
+                return isUnneccessary ? Unnecessary : Microsoft;
+            }
+
+            var customTagsBuilder = ImmutableArray.CreateBuilder<string>();
+            customTagsBuilder.AddRange(customTags.Concat(Microsoft));
+
+            if (!isConfigurable)
+            {
+                customTagsBuilder.Add(WellKnownDiagnosticTags.NotConfigurable);
+            }
+
+            if (isUnneccessary)
+            {
+                customTagsBuilder.Add(WellKnownDiagnosticTags.Unnecessary);
+            }
+
+            return customTagsBuilder.ToArray();
```

Comment:
there's ArrayBuilder, which also has ToArrayAndFree

Reply:
Ah, I always forget that as I keep switching back and forth between this repo and roslyn-analyzers, where we don't have ArrayBuilder. Will fix.

---
In reply to: [234867559](https://github.com/dotnet/roslyn/pull/30777#discussion_r234867559) [](ancestors = 234867559)

Reply:
So this code is in the CodeStyle layer, which does not have internals access to our pooled objects in Microsoft.CodeAnalysis. I am going to keep the current implementation and we can change this if we decide to link in the ArrayBuilder and other such utilities into CodeStyle layer. I will let @sharwell decide that.

---
In reply to: [235096083](https://github.com/dotnet/roslyn/pull/30777#discussion_r235096083) [](ancestors = 235096083,234867559)